### ISSUE TITLE
Choosing a test target using Stack

### ISSUE THREAD

**Issue by fatlazycat** [haskell/haskell-mode#1187]
Hi, probably my own lack of knowledge but I use haskell-session-change-target to set the target to be the test of my cabal file within a stack project.

When I start the repl it fails to find modules defined within the test target.

What obvious stupid thing am I doing wrong ? :)

Thanks

**Comment by geraldus** [comment]
@fatlazycat how do you start your REPL by the way?

**Comment by geraldus** [comment]
`haskell-session-change-target` also fails for me when I select executable.  What is the purpose of target selection, usually I'm either reloading file in REPL <kbd>C-c C-l</kbd> or running build command <kbd>C-c C-c</kbd>.

**Comment by fatlazycat** [comment]
In a test haskell file, i do haskell-session-change-target then start the repl ( I'm using spacemacs ) SPC m s b to load the current file into the repl which i believe is fn haskell-process-load-file

**Comment by fatlazycat** [comment]
Is the there a difference given I would start a stack repl with 
```
stack repl 'project name':test:'test target'
```
?

**Comment by fatlazycat** [comment]
Ah, can get it to work if I type that pattern when i do haskell-session-change-target so can work around it.

Guess the targets are just incorrect as it's just showing the cabal versions ?

**Comment by geraldus** [comment]
@gracjan can you help with this?  I have no much experience personally.  I guess target selecting makes some sense when running <kbd>C-c C-c</kbd> `haskell-process-cabal-build`.  I'm not sure.

**Comment by gracjan** [comment]
I tried to understand what you are talking about here but no avail. I'm lost. What is the problem again?

**Comment by fatlazycat** [comment]
Sorry, when showing the targets for ```haskell-session-change-target``` for a stack based project I see the cabal entries for executables & test suites. 

But when attempting to use any, such as the test target, it does not start CHCI correctly.

I assume it's running the command ```stack ghci test-suite```

But as far as I understand for a stack project you want to invoke it as 

```stack ghci project-name:test:test-suite```

**Comment by gracjan** [comment]
Ah, now it is clear as sun.

You are right that command line invocation is the same for cabal and stack projects, as I see it started to diverge. We have to split this functionality into two: cabal and stack.

Should not be that hard. @fatlazycat, are you able to look at this?

**Comment by fatlazycat** [comment]
Sure will try and take a look.

**Comment by gracjan** [comment]
Great!

**Comment by fatlazycat** [comment]
@gracjan from http://docs.haskellstack.org/en/stable/build_command/#target-syntax the targets have changed for build etc as well as ghci . Can you see any reason why I wouldn't change the targets it gets from the cabal file for haskell-cabal-enum-targets ?

**Comment by gracjan** [comment]
We still support plain `cabal` so we cannot drop old target syntax.

**Comment by fatlazycat** [comment]
Hi, yes just wondering if change the syntax when using stack ? Likely to break anything else ?

**Comment by gracjan** [comment]
Stack has its own documented syntax and we should be using that.

**Comment by csasarak** [comment]
Hi, I was bitten by this bug and it took me a while to make my way here. I had a look through the code to find where might be a good way to handle this, and I was thinking that maybe it's just the build target (not full command line) generation that needs to be modified at this point? I found the function `haskell-cabal-enum-targets` by looking at the definition of `haskell-session-change-target`. Would this be an appropriate place to start, or do the maintainers have a different design in mind?

**Comment by csasarak** [comment]
Hi everyone, we talked about breaking apart the code that runs build commands/ghci if possible a couple weeks ago here: https://github.com/haskell/haskell-mode/pull/1494

I made this issue to put down my thoughts/promote discussion around this. My thoughts are that I think we/I could create some set of structures which contain functions which is reponsible for calling the actual command. That'd put each of the commands for each tool/tool type in one spot and hopefully allow for users to define custom tool types if they wish by having a common interface for haskell tools. 

I also remember that as a new user I would get tripped up using stack since you can set haskell-process-type to 'stack-ghci, but then to do anything you'd have to also set a bunch of string commands to use it as well. This would hopefully solve that problem too. 

I was wondering what people think and if this is what commenters had in mind when they talked about "separating the code paths" for other tools?

**Comment by gracjan** [comment]
Good idea. That part of code and configuration is a mess that needs a cleanup.

I would focus on cabal/stack separation. Potential future tools can wait for their turn.

**Comment by csasarak** [comment]
Hi everyone, I'm working on this more. I also saw that there's some effort to combine haskell-process and haskell-session, I was wondering if this could affect that adversely. 

Best,
Chris

**Comment by gracjan** [comment]
Chris, you should be safe. Refactorings will take care to not step on bugfixes/features.
